### PR TITLE
Refactor jsonschema generator to use schemaview

### LINKED ISSUE #1033: Refactor jsonschema generator to use schemaview
Currently jsonschemagen uses the older restricted visitor pattern, which makes it harder to implement things like jsonschema composition, which are inherently recursive.

There are a few other things that it would be good to change, the order I would do this

- [ ] Extend the [compliance cases](https://github.com/linkml/linkml/blob/main/tests/test_generators/input/kitchen_sink_failtest_inst_01.yaml)
- [ ] make the existing tests more modular
- [ ] switch from visitor to schemaview
- [ ] avoid dict manipulation and use a model of json schema; also avoid string manipulation of the form `f"#/$defs/{...}`

The last one may be non-critical. However, I think some lightweight abstracting out of common patterns may be useful.

### PR DESCRIPTION
**Pull request by pkalita-lbl**
These changes address the four points outlined in #1033. They do not address the other issue mentioned in that one (#368); I think that would be better addressed separately. 

One extra change is here is upgrading to the 4.x series of the `jsonschema` package and including the `format` extras. This allows validating data instances against JSON schemas including things like `{ "format": "date-time" }` (which in turn results from LinkML schemas that include, for example, `range: xsddatetime`)

Fixes #1033

### CODE REVIEW

**Comment by pkalita-lbl** [comment]
I think this also resolves #606

File: linkml/generators/jsonschemagen.py [hunk 1]
```diff
@@ -32,7 +33,81 @@
     "xsdtime": ("string", "time"),
 }
 
-WITH_OPTIONAL_IDENTIFIER_SUFFIX = "__identifier_optional"
+class JsonSchema(UserDict):
```
Comment (cmungall) [inline linkml/generators/jsonschemagen.py:36]: I like how this is separated out

File: linkml/generators/jsonschemagen.py [hunk 2]
```diff
@@ -49,95 +124,71 @@ class JsonSchemaGenerator(Generator):
 
     # ClassVars
     generatorname = os.path.basename(__file__)
-    generatorversion = "0.0.2"
+    generatorversion = "0.0.3"
```
Comment (cmungall) [inline linkml/generators/jsonschemagen.py:127]: we haven't been doing this very consistently, need a strategy here. maybe overall version of linkml is good enough

File: linkml/generators/jsonschemagen.py [hunk 3]
```diff
@@ -49,95 +124,71 @@ class JsonSchemaGenerator(Generator):
...
     valid_formats = ["json"]
     visit_all_class_slots = True
     visit_all_slots = True
-    uses_schemaloader = True
-    #requires_metamodel = False
+    uses_schemaloader = False
```
Comment (cmungall) [inline linkml/generators/jsonschemagen.py:131]: 🍾

File: linkml/generators/jsonschemagen.py (hunk 2 above)
Comment (pkalita-lbl) [reply linkml/generators/jsonschemagen.py:128]: Okay yeah I wasn't sure. I saw it there so I figured I'd increment it. The `gen-json-schema` CLI is already hooked up to use the main LinkML version number when you pass the `--version` flag. I don't see any place the `generatorversion` variable is used within the `linkml` project. I'd be happy to just remove it unless you think there's a possibility of a downstream project using that variable.

File: linkml/generators/jsonschemagen.py [hunk 4]
```diff
@@ -154,136 +205,110 @@ def extract_permissible_text(pv):
...
         if (
             self.top_class is not None and camelcase(self.top_class) == camelcase(cls.name)
         ) or (self.top_class is None and cls.tree_root):
-            self.schemaobj.properties[underscore(aliased_slot_name)] = prop
+            self.top_level_schema.add_property(aliased_slot_name, prop, slot_is_required)
+
+    def serialize(self, **kwargs) -> str:
+        self.start_schema(**kwargs)
+        for enum_definition in self.schemaview.all_enums().values():
+            self.visit_enum(enum_definition)
```
Comment (cmungall) [inline linkml/generators/jsonschemagen.py:307]: it's maybe a bit odd keeping the method names as visit_X if it's not truly using the VP anymore... but YMMV

File: tests/test_generators/input/kitchen_sink_compliance_inst_01.yaml [hunk 5]
```diff
@@ -40,28 +37,25 @@
...
 - description: additional properties is closed
-  skip: false
   dataset:
     persons:
       - id: P
         name: john smith
         made_up_property: "abc"
 - description: optional open properties
-  skip: true
+  skip: not_closed value does not propagate to nested objects. By design?
```
Comment (cmungall) [inline tests/test_generators/input/kitchen_sink_compliance_inst_01.yaml:58]: hmm, perhaps Bill designed it that way, I can see the use case but a little unintuitive

File: tests/test_generators/input/kitchen_sink_compliance_inst_01.yaml [hunk 6]
```diff
@@ -1,37 +1,34 @@
+# By default these cases are expected to *not* validate. Use valid: true
```
Comment (cmungall) [inline tests/test_generators/input/kitchen_sink_compliance_inst_01.yaml:1]: `skip: true`?

Comment (cmungall) [reply tests/test_generators/input/kitchen_sink_compliance_inst_01.yaml:1]: ah wait I see the structure now

**Review by cmungall — APPROVED** [review]

File: linkml/generators/jsonschemagen.py (hunk 4 above)
Comment (pkalita-lbl) [reply linkml/generators/jsonschemagen.py:307]: I think you're right. I've started working on refactoring and adding methods for handling rules and I'll probably rename some existing methods as part of that.

**Pull request merged**